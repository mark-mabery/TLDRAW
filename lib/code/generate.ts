import Rectangle from "./rectangle"
import Circle from "./circle"
import Ellipse from "./ellipse"
import Polyline from "./polyline"
import Dot from "./dot"
import Ray from "./ray"
import Line from "./line"
import Vector from "./vector"
import Utils from "./utils"
import { codeShapes } from "./index"

const scope = {
  Dot,
  Circle,
  Ellipse,
  Ray,
  Line,
  Polyline,
  Rectangle,
  Vector,
  Utils,
}

/**
 * Evaluate code, collecting generated shapes in the shape set. Return the
 * collected shapes as an array.
 * @param code
 */
export function getShapesFromCode(code: string) {
  codeShapes.clear()

  new Function(...Object.keys(scope), `${code}`)(...Object.values(scope))

  const generatedShapes = Array.from(codeShapes.values()).map((instance) => {
    instance.shape.isGenerated = true
    return instance.shape
  })

  return generatedShapes
}
