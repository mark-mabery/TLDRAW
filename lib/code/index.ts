import { Shape } from "types"
import * as vec from "utils/vec"
import { getShapeUtils } from "lib/shapes"

export default class CodeShape<T extends Shape> {
  private _shape: T

  constructor(props: T) {
    this._shape = props
    shapeMap.add(this)
  }

  destroy() {
    shapeMap.delete(this)
  }

  moveTo(point: number[]) {
    this.shape.point = point
  }

  translate(delta: number[]) {
    this.shape.point = vec.add(this._shape.point, delta)
  }

  rotate(rotation: number) {
    this.shape.rotation = rotation
  }

  scale(scale: number) {
    return getShapeUtils(this.shape).scale(this.shape, scale)
  }

  getBounds() {
    return getShapeUtils(this.shape).getBounds(this.shape)
  }

  hitTest(point: number[]) {
    return getShapeUtils(this.shape).hitTest(this.shape, point)
  }

  get shape() {
    return this._shape
  }

  get point() {
    return [...this.shape.point]
  }

  get rotation() {
    return this.shape.rotation
  }
}

export const shapeMap = new Set<CodeShape<Shape>>([])
