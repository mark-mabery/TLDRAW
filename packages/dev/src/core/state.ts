import type {
  TLBinding,
  TLPage,
  TLPageState,
  TLPointerEventHandler,
  TLShapeChangeHandler,
} from '@tldraw/core'
import type { RectangleShape } from './box'
import type { LabelShape } from './label'
import { StateManager } from 'rko'

type Shapes = RectangleShape | LabelShape

interface State {
  page: TLPage<Shapes, TLBinding>
  pageState: TLPageState
  meta: {
    isDarkMode: boolean
  }
}

class AppState extends StateManager<State> {
  /* ----------------------- API ---------------------- */

  selectShape(shapeId: string) {
    this.patchState({
      pageState: {
        selectedIds: [shapeId],
      },
    })
  }

  deselect() {
    this.patchState({
      pageState: {
        selectedIds: [],
        editingId: undefined,
      },
    })
  }

  startEditingShape(shapeId: string) {
    this.patchState({
      pageState: {
        selectedIds: [shapeId],
        editingId: shapeId,
      },
    })
  }

  changeShapeText = (id: string, text: string) => {
    this.patchState({
      page: {
        shapes: {
          [id]: { text },
        },
      },
    })
  }

  /* --------------------- Events --------------------- */

  onPointCanvas: TLPointerEventHandler = (info) => {
    this.deselect()
  }

  onPointShape: TLPointerEventHandler = (info) => {
    this.selectShape(info.target)
  }

  onDoubleClickShape: TLPointerEventHandler = (info) => {
    this.startEditingShape(info.target)
  }

  onDoubleClickBounds: TLPointerEventHandler = (info) => {
    // Todo
  }

  onPointerDown: TLPointerEventHandler = (info) => {
    // Todo
  }

  onPointerUp: TLPointerEventHandler = (info) => {
    // Todo
  }

  onPointerMove: TLPointerEventHandler = (info) => {
    // Todo
  }

  onShapeChange: TLShapeChangeHandler<Shapes> = (shape) => {
    if (shape.type === 'rectangle' && shape.size) {
      this.patchState({
        page: {
          shapes: {
            [shape.id]: { ...shape, size: [...shape.size] },
          },
        },
      })
    }
  }
}

export const appState = new AppState({
  page: {
    id: 'page1',
    shapes: {
      rect1: {
        id: 'rect1',
        parentId: 'page1',
        name: 'Rectangle',
        childIndex: 1,
        type: 'rectangle',
        point: [0, 0],
        rotation: 0,
        size: [100, 100],
        text: 'Hello world!',
      },
      label1: {
        id: 'label1',
        parentId: 'page1',
        name: 'Label',
        childIndex: 2,
        type: 'label',
        point: [-200, -200],
        rotation: 0,
        text: 'My shape is stateful, I should still render while off screen!',
      },
      label2: {
        id: 'label2',
        parentId: 'page1',
        name: 'Label',
        childIndex: 2,
        type: 'label',
        point: [200, 200],
        rotation: 0,
        text: 'Hello world!',
      },
    },
    bindings: {},
  },
  pageState: {
    id: 'page1',
    selectedIds: [],
    camera: {
      point: [0, 0],
      zoom: 1,
    },
  },
  meta: {
    isDarkMode: false,
  },
})
