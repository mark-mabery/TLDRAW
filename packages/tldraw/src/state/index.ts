export * from './tlstate'
export * from './state-types'
