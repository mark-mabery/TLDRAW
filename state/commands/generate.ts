import Command from './command'
import history from '../history'
import { Data, Shape } from 'types'
import { deepClone, getPage, getShapes, setSelectedIds } from 'utils'

export default function generateCommand(
  data: Data,
  generatedShapes: Shape[]
): void {
  const initialShapes = getShapes(data)
    .filter((shape) => shape.isGenerated)
    .map(deepClone)

  history.execute(
    data,
    new Command({
      name: 'generate_shapes',
      category: 'canvas',
      do(data) {
        const { shapes } = getPage(data)
        initialShapes.forEach((shape) => delete shapes[shape.id])
        generatedShapes.forEach((shape) => (shapes[shape.id] = shape))
        setSelectedIds(data, [])
      },
      undo(data) {
        const { shapes } = getPage(data)
        generatedShapes.forEach((shape) => delete shapes[shape.id])
        initialShapes.forEach((shape) => (shapes[shape.id] = shape))
        setSelectedIds(data, [])
      },
    })
  )
}
