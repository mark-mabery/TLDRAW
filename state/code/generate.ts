import Rectangle from './rectangle'
import Ellipse from './ellipse'
import Polyline from './polyline'
import Dot from './dot'
import Ray from './ray'
import Line from './line'
import Arrow from './arrow'
import Draw from './draw'
import Text from './text'
import Utils from './utils'
import Vec from 'utils/vec'
import { NumberControl, VectorControl, codeControls, controls } from './control'
import { codeShapes } from './index'
import {
  CodeControl,
  Data,
  Shape,
  DashStyle,
  ColorStyle,
  FontSize,
  SizeStyle,
  CodeError,
} from 'types'
import { getPage, getShapes } from 'utils'
import { transform } from 'sucrase'
import { getErrorWithLineAndColumn, getFormattedCode } from 'utils/code'

const baseScope = {
  Dot,
  Ellipse,
  Ray,
  Line,
  Polyline,
  Rectangle,
  Vec,
  Utils,
  Arrow,
  Draw,
  Text,
  VectorControl,
  NumberControl,
  DashStyle,
  ColorStyle,
  SizeStyle,
  FontSize,
}

/**
 * Evaluate code, collecting generated shapes in the shape set. Return the
 * collected shapes as an array.
 * @param code
 */
export async function generateFromCode(
  data: Data,
  code: string
): Promise<{
  shapes: Shape[]
  controls: CodeControl[]
  error: CodeError
}> {
  codeControls.clear()
  codeShapes.clear()
  ;(window as any).isUpdatingCode = false
  ;(window as any).currentPageId = data.currentPageId

  const { currentPageId } = data
  const scope = { ...baseScope, controls, currentPageId }

  let generatedShapes: Shape[] = []
  let generatedControls: CodeControl[] = []
  let error: CodeError | null = null

  try {
    const formattedCode = getFormattedCode(code)

    const transformedCode = transform(formattedCode, {
      transforms: ['typescript'],
    })?.code

    new Function(...Object.keys(scope), `${transformedCode}`)(
      ...Object.values(scope)
    )

    const startingChildIndex =
      getShapes(data)
        .filter((shape) => shape.parentId === data.currentPageId)
        .sort((a, b) => a.childIndex - b.childIndex)[0]?.childIndex || 1

    generatedShapes = Array.from(codeShapes.values())
      .sort((a, b) => a.shape.childIndex - b.shape.childIndex)
      .map((instance, i) => ({
        ...instance.shape,
        isGenerated: true,
        parentId: getPage(data).id,
        childIndex: startingChildIndex + i,
      }))

    generatedControls = Array.from(codeControls.values())
  } catch (e) {
    error = getErrorWithLineAndColumn(e)
  }

  return { shapes: generatedShapes, controls: generatedControls, error }
}

/**
 * Evaluate code, collecting generated shapes in the shape set. Return the
 * collected shapes as an array.
 * @param code
 */
export async function updateFromCode(
  data: Data,
  code: string
): Promise<{
  shapes: Shape[]
}> {
  codeShapes.clear()
  ;(window as any).isUpdatingCode = true
  ;(window as any).currentPageId = data.currentPageId

  const { currentPageId } = data

  const newControls = Object.fromEntries(
    Object.entries(data.codeControls).map(([_, control]) => [
      control.label,
      control.value,
    ])
  )

  const scope = {
    ...baseScope,
    currentPageId,
    controls: newControls,
  }

  const startingChildIndex =
    getShapes(data)
      .filter((shape) => shape.parentId === data.currentPageId)
      .sort((a, b) => a.childIndex - b.childIndex)[0]?.childIndex || 1

  const transformed = transform(code, {
    transforms: ['typescript'],
  }).code

  new Function(...Object.keys(scope), `${transformed}`)(...Object.values(scope))

  const generatedShapes = Array.from(codeShapes.values())
    .sort((a, b) => a.shape.childIndex - b.shape.childIndex)
    .map((instance, i) => ({
      ...instance.shape,
      isGenerated: true,
      parentId: getPage(data).id,
      childIndex: startingChildIndex + i,
    }))

  return { shapes: generatedShapes }
}
