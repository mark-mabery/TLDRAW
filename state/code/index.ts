import { Mutable, Shape, ShapeUtility } from 'types'
import { createShape, getShapeUtils } from 'state/shape-utils'
import vec from 'utils/vec'
import Vector from './vector'
import Utils from './utils'

export const codeShapes = new Set<CodeShape<Shape>>([])

/**
 * A base class for code shapes. Note that creating a shape adds it to the
 * shape map, while deleting it removes it from the collected shapes set
 */
export default class CodeShape<T extends Shape> {
  private _shape: Mutable<T>
  private utils: ShapeUtility<T>

  constructor(props: T) {
    this._shape = createShape(props.type, props) as Mutable<T>
    this.utils = getShapeUtils<T>(this._shape)
    codeShapes.add(this)
  }

  destroy(): void {
    codeShapes.delete(this)
  }

  moveTo(point: Vector): CodeShape<T> {
    this.utils.setProperty(this._shape, 'point', Utils.vectorToPoint(point))
    return this
  }

  translate(delta: Vector): CodeShape<T> {
    this.utils.setProperty(
      this._shape,
      'point',
      vec.add(this._shape.point, Utils.vectorToPoint(delta))
    )
    return this
  }

  rotate(rotation: number): CodeShape<T> {
    this.utils.setProperty(this._shape, 'rotation', rotation)
    return this
  }

  getBounds(): CodeShape<T> {
    this.utils.getBounds(this.shape)
    return this
  }

  hitTest(point: Vector): CodeShape<T> {
    this.utils.hitTest(this.shape, Utils.vectorToPoint(point))
    return this
  }

  get shape(): T {
    return this._shape
  }

  get point(): number[] {
    return [...this.shape.point]
  }

  get rotation(): number {
    return this.shape.rotation
  }
}
