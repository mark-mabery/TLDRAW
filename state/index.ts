import state, { useSelector } from "./state"

export default state

export { useSelector }
