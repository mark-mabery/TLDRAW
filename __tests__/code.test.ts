import state from 'state'
import { generateFromCode } from 'state/code/generate'
import * as json from './__mocks__/document.json'
import tld from 'utils/tld'

jest.useRealTimers()

state.reset()
state.send('MOUNTED').send('LOADED_FROM_FILE', { json: JSON.stringify(json) })
state.send('CLEARED_PAGE')

describe('selection', () => {
  it('opens and closes the code panel', () => {
    expect(state.data.settings.isCodeOpen).toBe(false)
    state.send('TOGGLED_CODE_PANEL_OPEN')
    expect(state.data.settings.isCodeOpen).toBe(true)
    state.send('TOGGLED_CODE_PANEL_OPEN')
    expect(state.data.settings.isCodeOpen).toBe(false)
  })

  it('saves changes to code', () => {
    expect(tld.getShapes(state.data).length).toBe(0)

    const code = `// hello world!`

    state.send('SAVED_CODE', { code })

    expect(state.data.document.code[state.data.currentCodeFileId].code).toBe(
      code
    )
  })

  it('generates shapes', async () => {
    const code = `
    const rectangle = new Rectangle({
      id: "test-rectangle",
      name: 'Test Rectangle',
      point: [100, 100],
      size: [200, 200],
      style: {
        size: SizeStyle.Medium,
        color: ColorStyle.Red,
        dash: DashStyle.Dotted,
      },
    })
    `

    const { controls, shapes } = await generateFromCode(state.data, code)

    state.send('GENERATED_FROM_CODE', { controls, shapes })

    expect(tld.getShapes(state.data)).toMatchSnapshot(
      'generated rectangle from code'
    )
  })

  it('creates a code control', async () => {
    const code = `
    new NumberControl({
      id: "test-number-control",
      label: "x"
    })
    `

    const { controls, shapes } = await generateFromCode(state.data, code)

    state.send('GENERATED_FROM_CODE', { controls, shapes })

    expect(state.data.codeControls).toMatchSnapshot(
      'generated code controls from code'
    )
  })

  it('updates a code control', async () => {
    const code = `
    new NumberControl({
      id: "test-number-control",
      label: "x"
    })

    new VectorControl({
      id: "test-vector-control",
      label: "size"
    })

    const rectangle = new Rectangle({
      id: "test-rectangle",
      name: 'Test Rectangle',
      point: [controls.x, 100],
      size: controls.size,
      style: {
        size: SizeStyle.Medium,
        color: ColorStyle.Red,
        dash: DashStyle.Dotted,
      },
    })
    `

    const { controls, shapes } = await generateFromCode(state.data, code)

    state.send('GENERATED_FROM_CODE', { controls, shapes })

    state.send('CHANGED_CODE_CONTROL', { 'test-number-control': 100 })

    expect(state.data.codeControls).toMatchSnapshot(
      'data in state after changing control'
    )

    expect(tld.getShape(state.data, 'test-rectangle')).toMatchSnapshot(
      'rectangle in state after changing code control'
    )
  })

  /* -------------------- Readonly -------------------- */

  it('does not saves changes to code when readonly', () => {
    state.send('CLEARED_PAGE')

    expect(tld.getShapes(state.data).length).toBe(0)

    const code = `// hello world!`

    state
      .send('SAVED_CODE', { code })
      .send('TOGGLED_READ_ONLY')
      .send('SAVED_CODE', { code: '' })

    expect(state.data.document.code[state.data.currentCodeFileId].code).toBe(
      code
    )

    state.send('TOGGLED_READ_ONLY').send('SAVED_CODE', { code: '' })

    expect(state.data.document.code[state.data.currentCodeFileId].code).toBe('')
  })

  /* --------------------- Methods -------------------- */

  it('moves shape to front', async () => {
    null
  })

  it('moves shape forward', async () => {
    null
  })

  it('moves shape backward', async () => {
    null
  })

  it('moves shape to back', async () => {
    null
  })

  it('rotates a shape', async () => {
    null
  })

  it('rotates a shape by a delta', async () => {
    null
  })

  it('translates a shape', async () => {
    null
  })

  it('translates a shape by a delta', async () => {
    null
  })

  /* --------------------- Shapes --------------------- */

  it('generates a rectangle shape', async () => {
    state.send('CLEARED_PAGE')
    const code = `
    const rectangle = new Rectangle({
      id: "test-rectangle",
      name: 'Test Rectangle',
      point: [100, 100],
      size: [200, 200],
      style: {
        size: SizeStyle.Medium,
        color: ColorStyle.Red,
        dash: DashStyle.Dotted,
      },
    })
    `

    const { controls, shapes } = await generateFromCode(state.data, code)

    state.send('GENERATED_FROM_CODE', { controls, shapes })

    expect(tld.getShapes(state.data)).toMatchSnapshot(
      'generated rectangle from code'
    )
  })

  it('changes a rectangle size', async () => {
    null
  })

  it('generates an ellipse shape', async () => {
    state.send('CLEARED_PAGE')
    const code = `
    const ellipse = new Ellipse({
      id: 'test-ellipse',
      name: 'Test ellipse',
      point: [100, 100],
      radiusX: 100,
      radiusY: 200,
      style: {
        size: SizeStyle.Medium,
        color: ColorStyle.Red,
        dash: DashStyle.Dotted,
      },
    })
    `

    const { controls, shapes } = await generateFromCode(state.data, code)

    state.send('GENERATED_FROM_CODE', { controls, shapes })

    expect(tld.getShapes(state.data)).toMatchSnapshot(
      'generated ellipse from code'
    )
  })

  it('generates a draw shape', async () => {
    state.send('CLEARED_PAGE')
    const code = `
    const ellipse = new Draw({
      id: 'test-draw',
      name: 'Test draw',
      points: [[100, 100], [200,200], [300,300]],
      style: {
        size: SizeStyle.Medium,
        color: ColorStyle.Red,
        dash: DashStyle.Dotted,
      },
    })
    `

    const { controls, shapes } = await generateFromCode(state.data, code)

    state.send('GENERATED_FROM_CODE', { controls, shapes })

    expect(tld.getShapes(state.data)).toMatchSnapshot(
      'generated draw from code'
    )
  })

  it('generates an arrow shape', async () => {
    state.send('CLEARED_PAGE')
    const code = `
    const draw = new Arrow({
      id: 'test-draw',
      name: 'Test draw',
      points: [[100, 100], [200,200], [300,300]],
      style: {
        size: SizeStyle.Medium,
        color: ColorStyle.Red,
        dash: DashStyle.Dotted,
      },
    })
    `

    const { controls, shapes } = await generateFromCode(state.data, code)

    state.send('GENERATED_FROM_CODE', { controls, shapes })

    expect(tld.getShapes(state.data)).toMatchSnapshot(
      'generated draw from code'
    )
  })

  it('generates a text shape', async () => {
    state.send('CLEARED_PAGE')
    const code = `
    const text = new Text({
      id: 'test-text',
      name: 'Test text',
      point: [100, 100],
      text: 'Hello world!',
      style: {
        size: SizeStyle.Large,
        color: ColorStyle.Red,
        dash: DashStyle.Dotted,
      },
    })
    `

    const { controls, shapes } = await generateFromCode(state.data, code)

    state.send('GENERATED_FROM_CODE', { controls, shapes })

    expect(tld.getShapes(state.data)).toMatchSnapshot(
      'generated draw from code'
    )
  })
})
