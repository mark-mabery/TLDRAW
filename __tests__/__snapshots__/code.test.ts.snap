// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selection creates a code control: generated code controls from code 1`] = `
Object {
  "test-number-control": Object {
    "id": "test-number-control",
    "label": "x",
    "step": 1,
    "type": "number",
    "value": 0,
  },
}
`;

exports[`selection generates a draw shape: generated draw from code 1`] = `
Array [
  Object {
    "childIndex": 1,
    "id": "test-draw",
    "isGenerated": true,
    "name": "Test draw",
    "parentId": "page1",
    "point": Array [
      0,
      0,
    ],
    "points": Array [
      Array [
        100,
        100,
      ],
      Array [
        200,
        200,
      ],
      Array [
        300,
        300,
      ],
    ],
    "rotation": 0,
    "style": Object {
      "color": "Red",
      "dash": "Dotted",
      "isFilled": false,
      "size": "Medium",
    },
    "type": "draw",
  },
]
`;

exports[`selection generates a rectangle shape: generated rectangle from code 1`] = `
Array [
  Object {
    "childIndex": 1,
    "id": "test-rectangle",
    "isGenerated": true,
    "name": "Test Rectangle",
    "parentId": "page1",
    "point": Array [
      100,
      100,
    ],
    "radius": 2,
    "rotation": 0,
    "size": Array [
      200,
      200,
    ],
    "style": Object {
      "color": "Red",
      "dash": "Dotted",
      "isFilled": false,
      "size": "Medium",
    },
    "type": "rectangle",
  },
]
`;

exports[`selection generates a text shape: generated draw from code 1`] = `
Array [
  Object {
    "childIndex": 1,
    "id": "test-text",
    "isGenerated": true,
    "name": "Test text",
    "parentId": "page1",
    "point": Array [
      100,
      100,
    ],
    "rotation": 0,
    "scale": 1,
    "style": Object {
      "color": "Red",
      "dash": "Dotted",
      "isFilled": false,
      "size": "Large",
    },
    "text": "Hello world!",
    "type": "text",
  },
]
`;

exports[`selection generates an arrow shape: generated draw from code 1`] = `
Array [
  Object {
    "bend": 0,
    "childIndex": 1,
    "decorations": Object {
      "end": "Arrow",
      "middle": null,
      "start": null,
    },
    "handles": Object {
      "bend": Object {
        "id": "bend",
        "index": 2,
        "point": Array [
          50,
          50,
        ],
      },
      "end": Object {
        "id": "end",
        "index": 1,
        "point": Array [
          100,
          100,
        ],
      },
      "start": Object {
        "id": "start",
        "index": 0,
        "point": Array [
          0,
          0,
        ],
      },
    },
    "id": "test-draw",
    "isGenerated": true,
    "name": "Test draw",
    "parentId": "page1",
    "point": Array [
      0,
      0,
    ],
    "points": Array [
      Array [
        100,
        100,
      ],
      Array [
        200,
        200,
      ],
      Array [
        300,
        300,
      ],
    ],
    "rotation": 0,
    "style": Object {
      "color": "Red",
      "dash": "Dotted",
      "isFilled": false,
      "size": "Medium",
    },
    "type": "arrow",
  },
]
`;

exports[`selection generates an ellipse shape: generated ellipse from code 1`] = `
Array [
  Object {
    "childIndex": 1,
    "id": "test-ellipse",
    "isGenerated": true,
    "name": "Test ellipse",
    "parentId": "page1",
    "point": Array [
      100,
      100,
    ],
    "radiusX": 100,
    "radiusY": 200,
    "rotation": 0,
    "style": Object {
      "color": "Red",
      "dash": "Dotted",
      "isFilled": false,
      "size": "Medium",
    },
    "type": "ellipse",
  },
]
`;

exports[`selection generates shapes: generated rectangle from code 1`] = `
Array [
  Object {
    "childIndex": 1,
    "id": "test-rectangle",
    "isGenerated": true,
    "name": "Test Rectangle",
    "parentId": "page1",
    "point": Array [
      100,
      100,
    ],
    "radius": 2,
    "rotation": 0,
    "size": Array [
      200,
      200,
    ],
    "style": Object {
      "color": "Red",
      "dash": "Dotted",
      "isFilled": false,
      "size": "Medium",
    },
    "type": "rectangle",
  },
]
`;

exports[`selection updates a code control: data in state after changing control 1`] = `
Object {
  "test-number-control": Object {
    "id": "test-number-control",
    "label": "x",
    "step": 1,
    "type": "number",
    "value": 100,
  },
  "test-vector-control": Object {
    "id": "test-vector-control",
    "isNormalized": false,
    "label": "size",
    "type": "vector",
    "value": Array [
      0,
      0,
    ],
  },
}
`;

exports[`selection updates a code control: rectangle in state after changing code control 1`] = `
Object {
  "childIndex": 1,
  "id": "test-rectangle",
  "isGenerated": true,
  "name": "Test Rectangle",
  "parentId": "page1",
  "point": Array [
    0,
    100,
  ],
  "radius": 2,
  "rotation": 0,
  "size": Array [
    0,
    0,
  ],
  "style": Object {
    "color": "Red",
    "dash": "Dotted",
    "isFilled": false,
    "size": "Medium",
  },
  "type": "rectangle",
}
`;
